rig/routing_table/entries.py

```diff
@@ -23,7 +24,18 @@ class RoutingTableEntry(namedtuple("RoutingTableEntry", "route key mask")):
     mask : int
         32-bit unsigned integer mask to apply to keys of packets arriving at
         the router.
+    sources : {:py:class:`~.Routes`, ...}
+        Directions in which a packet may enter the router before taking this
+        route. This is opposite to the link used (e.g., ``north`` in this set
+        indicates that a packet would enter on the ``south`` link). If the
+        `sources` and `route` are equivalent and contain only one item then it
+        *might* be possible to remove the entry and instead use default
+        routing.
     """
+    def __new__(cls, route, key, mask, sources={None}):
+        return super(RoutingTableEntry, cls).__new__(
+                cls, route, key, mask, set(sources))
+
     def __str__(self):
         """Get an easily readable representation of the routing table entry.
```

Comment:
Could probably do with being updated to the form "XXX...XXX -> N NE ..." when the source is {None} and "S SW -> XXX....XXX -> N NE..." when it isn't (not-showing the None in the source list, if present).